### ISSUE TITLE
pt-archiver style iteration

### ISSUE THREAD

**Issue by adamchainz** [adamchainz/django-mysql#17]
It may be useful for `QuerySet`s to support the same iteration strategy as pt-archiver for archiving jobs - pulling rows one at a time and committing their update/delete in chunks.

**Comment by Geekfish** [comment]
Hi @adamchainz 

I know it's a bit of an old ticket, but I'd be interested in working on this.
Would you have any suggestions on a setup to test the efficiency of difference solutions outside of unit tests?

Many thanks for your great work!

**Comment by adamchainz** [comment]
I think you really need a big table with realistic fragmentation to investigate this properly. Also look in the source code for `pt-table-sync`.

Are you having any problems with the current algorithm?

**Comment by Geekfish** [comment]
We ran into some inefficiency when dealing with queries that have sparse results over big tables. This was due to the query conditions but also due to removed rows (with ids earlier in the table being more sparse for example).
We have found some workarounds for our cases (ex. not using chunks at all for queries with smaller counts) and the iterator still works nicely most of the time, but I was personally interested in a better solution for sparse distributions.

**Comment by adamchainz** [comment]
Yeah sparse results are difficult because there's no predicting when the density of results changes. What the current algorithm does is just react to this by decreasing the chunk size rapidly.
